File: android/core/src/main/java/com/ghamza/janus/core/JaHandle.kt
package com.ghamza.janus.core

import com.ghamza.janus.bindings.Handle
import com.ghamza.janus.bindings.HandleCallback
import com.ghamza.janus.bindings.Jsep
import kotlinx.coroutines.channels.ProducerScope
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.callbackFlow
import java.time.Duration

class JaHandle(val handle: Handle): HandleCallback {
    private var events: ProducerScope<String>? = null

    fun stream() = callbackFlow {
        events = this
        handle.startEventLoop(this@JaHandle)
        awaitClose { }
    }

    suspend fun fireAndForget(msg: String) {
        handle.fireAndForget(msg)
    }

    suspend fun fireAndForget(msg: String, jsep: Jsep) {
        handle.fireAndForgetWithJsep(msg, jsep)
    }

    suspend fun sendWaitOnAck(msg: String, timeout: Duration) {
        handle.sendWaitonAck(msg, timeout)
    }

    suspend fun sendWaitOnResult(msg: String, timeout: Duration): String {
        return handle.sendWaitonResult(msg, timeout)
    }

    override fun onEvent(event: String) {
        events?.trySend(event)
    }
}

File: apple/Sources/JanusGateway/JaHandle.swift
//
//  JaHandle.swift
//
//
//  Created by Hamza Jadid on 16/09/2024.
//

import Foundation
import UniFFI

/// General purpose plugin handle
public final class JaHandle {
    let handle: Handle
    public var delegate: JaHandleDelegate?
    private var continuation: AsyncStream<String>.Continuation?

    /// Get an async stream of incoming Janus events for this handle
    ///
    /// - Returns: An async stream of incoming events
    public var events: AsyncStream<String> {
        get async {
            await handle.startEventLoop(cb: self)

            return AsyncStream { continuation in
                self.continuation = continuation
            }
        }
    }

    init(handle: Handle) {
        self.handle = handle
    }

    /// Sends a message without waiting for any response or acknowledgment
    ///
    /// - Parameters:
    ///     - msg: Message to be sent
    public func fireAndForget(msg: String) async throws {
        try await handle.fireAndForget(msg: msg)
    }

    /// Sends a message without waiting for any response or acknowledgment
    ///
    /// - Parameters:
    ///     - msg: Message to be sent
    ///     - jsep: JavaScript Session Establishment Protocol
    public func fireAndForget(msg: String, jsep: Jsep) async throws {
        try await handle.fireAndForgetWithJsep(msg: msg, jsep: jsep)
    }

    /// Sends a message and waits until the server acknowledges or timeout
    ///
    /// - Parameters:
    ///     - msg: Message to be sent
    ///     - timeout: The maximum amount of time to wait on an acknowledgment before we consider
    ///     the request as failed or times out.
    public func sendWaitonAck(
        msg: String, timeout: TimeInterval
    ) async throws {
        try await handle.sendWaitonAck(msg: msg, timeout: timeout)
    }

    /// Sends a message and waits until the server returns a response or timeout
    ///
    /// - Parameters:
    ///     - msg: Message to be sent
    ///     - timeout: The maximum amount of time to wait on a response before we consider the
    ///     request as failed or times out.
    public func sendWaitonResult(
        msg: String,
        timeout: TimeInterval
    ) async throws -> String {
        try await handle.sendWaitonResult(msg: msg, timeout: timeout)
    }
}

extension JaHandle: HandleCallback {
    public func onEvent(event: String) {
        delegate?.didReceive(event: event)
        continuation?.yield(event)
    }
}
